upgrades/steps_28.go

```diff
@@ -139,59 +156,163 @@ var getUpgradeStepsClient = func(caller base.APICaller) UpgradeStepsClient {
-			return nil, errors.Annotatef(err, "failed to unmarshal uniter state from %q", uniterStateFile)
+		storageSt, filename, err := readStorageState(uniterStateDir)
+		if err != nil && !os.IsNotExist(err) {
+			return nil, err
+		}
+		if storageSt != nil {
+			fileNames.Add(filename)
+			uniterStates[i].StorageState = storageSt
 		}
-		uniterStates[tag] = string(data)
 	}
 
+	logger.Criticalf("filenames %d %s", fileNames.Size(), fileNames.Values())
 	// No state files, nothing to do.
-	if len(fileNames) == 0 {
+	if fileNames.IsEmpty() {
 		return nil, nil
 	}
 
 	client := getUpgradeStepsClient(ctx.APIState())
 	if err := client.WriteUniterState(uniterStates); err != nil {
 		return nil, errors.Annotatef(err, "unable to set state for units %q", strings.Join(unitNames, ", "))
 	}
-	return fileNames, nil
+	return fileNames.Values(), nil
+}
+
+func readUniterState(uniterStateDir string) (*string, string, error) {
+	uniterStateFile := filepath.Join(uniterStateDir, "uniter")
+	// Read into a State as we can validate the data.
+	var st operation.State
+	if err := utils.ReadYaml(uniterStateFile, &st); err != nil {
+		// No file available, move on.
+		if os.IsNotExist(err) {
+			return nil, "", err
+		}
+		return nil, "", errors.Annotatef(err, "failed to read uniter state from %q", uniterStateFile)
+	}
+	if err := st.Validate(); err != nil {
+		return nil, "", errors.Annotatef(err, "failed to validate uniter state from %q", uniterStateFile)
+	}
+
+	data, err := yaml.Marshal(st)
+	if err != nil {
+		return nil, "", errors.Annotatef(err, "failed to unmarshal uniter state from %q", uniterStateFile)
+	}
+	str := string(data)
+
+	return &str, uniterStateFile, nil
 }
 
 func caasOperatorLocalState(ctx Context, unitNames []string) ([]string, error) {
 	// TODO: (hml) 27-Mar-2020.
 	// Implement when relations, storage, metrics and the operator state are moved.
 	return nil, nil
 }
+
+func readStorageState(uniterStateDir string) (*string, string, error) {
+	storageStateDir := filepath.Join(uniterStateDir, "storage")
+	storage, err := readAllStateFiles(storageStateDir)
+	if err != nil {
+		return nil, "", errors.Annotatef(err, "cannot load storage state from %q", storageStateDir)
+	}
+	if len(storage) < 1 {
+		return nil, "", nil
+	}
+
+	data, err := yaml.Marshal(storage)
+	if err != nil {
+		return nil, "", errors.Annotatef(err, "failed to unmarshal storage state from %q", storageStateDir)
+	}
+	str := string(data)
+
+	return &str, storageStateDir, nil
+}
+
+// readAllStateFiles loads and returns every stateFile persisted inside
+func readAllStateFiles(dirPath string) (map[string]bool, error) {
+	if _, err := os.Stat(dirPath); os.IsNotExist(err) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	fis, err := ioutil.ReadDir(dirPath)
+	if err != nil {
+		return nil, err
+	}
+	files := make(map[string]bool)
+	for _, fi := range fis {
+		if fi.IsDir() {
+			continue
+		}
+		storageId := fi.Name()
+		if i := strings.LastIndex(storageId, "-"); i > 0 {
```

Comment:
If you want, you could make this read a bit cleaner and shorter if you calculate `i` before the `if` and change it to `if i <= 0 { continue }`

Reply:
done